// spinalgraph/build/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SpinalGraph", {
  enumerable: true,
  get: function get() {
    return _SpinalGraph.default;
  }
});
Object.defineProperty(exports, "SpinalNode", {
  enumerable: true,
  get: function get() {
    return _SpinalNode.default;
  }
});
Object.defineProperty(exports, "SpinalContext", {
  enumerable: true,
  get: function get() {
    return _SpinalContext.default;
  }
});
Object.defineProperty(exports, "SpinalRelationRef", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationRef.default;
  }
});
Object.defineProperty(exports, "SpinalRelationLstPtr", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationLstPtr.default;
  }
});
Object.defineProperty(exports, "SpinalRelationPtrLst", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationPtrLst.default;
  }
});
Object.defineProperty(exports, "SpinalRelationFactory", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationFactory.SpinalRelationFactory;
  }
});
Object.defineProperty(exports, "SPINAL_RELATION_TYPE", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationFactory.SPINAL_RELATION_TYPE;
  }
});
Object.defineProperty(exports, "SPINAL_RELATION_LST_PTR_TYPE", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationFactory.SPINAL_RELATION_LST_PTR_TYPE;
  }
});
Object.defineProperty(exports, "SPINAL_RELATION_PTR_LST_TYPE", {
  enumerable: true,
  get: function get() {
    return _SpinalRelationFactory.SPINAL_RELATION_PTR_LST_TYPE;
  }
});
exports.GraphFunction = void 0;

var _SpinalGraph = _interopRequireDefault(require("./Nodes/SpinalGraph"));

var _SpinalNode = _interopRequireDefault(require("./Nodes/SpinalNode"));

var _SpinalContext = _interopRequireDefault(require("./Nodes/SpinalContext"));

var _SpinalRelationRef = _interopRequireDefault(require("./Relations/SpinalRelationRef"));

var _SpinalRelationLstPtr = _interopRequireDefault(require("./Relations/SpinalRelationLstPtr"));

var _SpinalRelationPtrLst = _interopRequireDefault(require("./Relations/SpinalRelationPtrLst"));

var GraphFunction = _interopRequireWildcard(require("./GraphFunctionsLib/GraphFunctions"));

exports.GraphFunction = GraphFunction;

var _SpinalRelationFactory = require("./Relations/SpinalRelationFactory");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
