// spinalgraph/build/Nodes/SpinalContext.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _SpinalNode = _interopRequireDefault(require("./SpinalNode"));

var _spinalCoreConnectorjs = _interopRequireDefault(require("spinal-core-connectorjs"));

var _SpinalRelationFactory = require("../Relations/SpinalRelationFactory");

var _Utilities = require("../Utilities");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const globalType = typeof window === "undefined" ? global : window;

class SpinalContext extends _SpinalNode.default {
  /**
   *
   * @param type of the context
   * @param name of the context
   * @param element of the node
   */
  constructor(type = "SpinalContext", name = "undefined", element = new globalType.Model()) {
    super(type, element);
    this.add_attr({
      relationIds: new globalType.Lst(),
      relationsNames: new globalType.Lst()
    });
    this.info.id.set((0, _Utilities.guid)(this.constructor.name));
    this.info.add_attr({
      name: typeof name === "undefined" ? this.info.id.get() : name
    });
  }
  /**
   * Create a relation between a node and another node/model. this function can be used to linked two context
   * @param parent {SpinalNode} parent of the relation
   * @param child {SpinalNode | Model} child of the relation
   * @param relationName {string} name of the relation
   * @return {Str}
   */


  createRelation(parent, child, relationName) {
    if (child instanceof _SpinalNode.default || child instanceof globalType.Model) {
      const relation = _SpinalRelationFactory.SpinalRelationFactory.getNewRelation(relationName, _SpinalRelationFactory.SPINAL_RELATION_PTR_LST_TYPE);

      relation.setParent(parent);
      relation.addChild(child);
      this.addRelationId(relation.id);
      return relation.id;
    } //Todo create Some Custom Error


    throw new Error("Cannot add a child witch is not an instance of SpinalNode or Model.");
  }
  /**
   * Add the relation id to the relation ids displayable by this context
   * @param relationId {string}
   */


  addRelationId(relationId) {
    this.relationIds.push(relationId);
  }
  /**
   * Remove all the nodes associated to this context from the graph
   */


  removeFromGraph() {} //Todo use DFS to remove all the context from the graph

  /**
   * Add Child to the context with a spinalRelationLstPtrType
   * @param child {SpinalNode| Model} to be added as child
   * @param relationName {string} name of the relation
   * @param relationType {int} this parameter is here only to properly override the parent method.
   */


  addChild(child, relationName, relationType = _SpinalRelationFactory.SPINAL_RELATION_PTR_LST_TYPE) {
    super.addChild(child, relationName, _SpinalRelationFactory.SPINAL_RELATION_PTR_LST_TYPE);
  }

}

_spinalCoreConnectorjs.default.register_models([SpinalContext]);

var _default = SpinalContext;
exports.default = _default;
