// spinalgraph/build/Relations/BaseSpinalRelation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _spinalCoreConnectorjs = _interopRequireDefault(require("spinal-core-connectorjs"));

var _Utilities = require("../Utilities");

var _SpinalNode = _interopRequireDefault(require("../Nodes/SpinalNode"));

var _SpinalNodePointer = _interopRequireDefault(require("../SpinalNodePointer"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const globalType = typeof window === "undefined" ? global : window;

class BaseSpinalRelation extends globalType.Model {
  constructor(name) {
    super();
    this.add_attr({
      id: (0, _Utilities.guid)(name),
      name: name,
      parent: new _SpinalNodePointer.default()
    });
  }
  /**
   * Return the parent of the relation.
   * @return {Promise} return a promise where the resolve function parameter is the parent
   */


  getParent() {
    return (0, _Utilities.promiseLoad)(this.parent);
  }
  /**
   * setElement the parent of the relation. If a parent was already setElement the parent relation is removed.
   * @param parent {SpinalNode}
   */


  setParent(parent) {
    if (typeof parent !== "undefined" && parent instanceof _SpinalNode.default) this.parent.setElement(parent);
  }
  /**
   * Returns the name of the relation
   * @return {Str} Name of the relation
   */


  getName() {
    return this.name;
  }
  /**
   * Returns the type of the relation
   * @return {Number} Type of the relation
   */


  getType() {
    return -1;
  }
  /**
   * Return all the children of the relation
   * @return {Promise<globalType.Lst>}
   */


  getChildren() {
    // noinspection JSValidateTypes
    return Promise.resolve();
  }
  /**
   * Add node as child of the relation if node is a model create a node.
   * @param node {SpinalNode | globalType.Model}
   */


  addChild(node) {}
  /**
   * This function retrieves all the id from children of the relation and return them inside an array.
   * @return {Array} containing all the children Id of the relation
   */


  getChildrenIds() {}
  /**
   * Remove the child from the relation
   * @param node {SpinalNode} child of the relation
   * @return {Promise<boolean>}
   */


  removeChild(node) {
    return _asyncToGenerator(function* () {})();
  }
  /**
   * Removes all children from the relation
   */


  removeChildren() {
    this.getChildren().then(children => {
      for (let i = 0; i < children.length; i++) {
        this.removeChild(children[i]).then();
      }
    }).catch(e => {
      console.error("cannot remove child ", e);
    });
  }
  /**
   * Removes the relation from the graph
   */


  removeFromGraph() {
    this._removeFromParent();

    this.removeChildren();
  }
  /**
   * Removes the relation from the parent
   * @private
   */


  _removeFromParent() {
    this.getParent().then(parentNode => {
      let relationMap = parentNode._getRelationListType(this.getType());

      relationMap.delete(this.name.get());
    });
  }

}

_spinalCoreConnectorjs.default.register_models([BaseSpinalRelation]);

var _default = BaseSpinalRelation;
exports.default = _default;
