// spinalgraph/build/Relations/SpinalRelationLstPtr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _BaseSpinalRelation = _interopRequireDefault(require("./BaseSpinalRelation"));

var _SpinalRelationFactory = require("./SpinalRelationFactory");

var _SpinalNode = _interopRequireDefault(require("../Nodes/SpinalNode"));

var _Utilities = require("../Utilities");

var _SpinalNodePointer = _interopRequireDefault(require("../SpinalNodePointer"));

var _spinalCoreConnectorjs = _interopRequireDefault(require("spinal-core-connectorjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const globalType = typeof window === "undefined" ? global : window;

class SpinalRelationLstPtr extends _BaseSpinalRelation.default {
  constructor(name) {
    super(name);
    this.add_attr({
      children: new globalType.Lst()
    });
  }
  /**
   * This function retrieve all the id from children of the relation and return them inside an array.
   * @return {Array} containing all the children Id of the relation
   */


  getChildrenIds() {
    const res = [];

    for (let i = 0; i < this.children.length; i++) {
      res.push(this.children[i].getId());
    }

    return res;
  }
  /**
   * Return all the children of the relation
   * @return {Promise<globalType.Lst>}
   */


  getChildren() {
    var _this = this;

    return _asyncToGenerator(function* () {
      const promiseList = [];

      for (let i = 0; i < _this.children.length; i++) {
        let ptr = _this.children[i];
        promiseList.push((0, _Utilities.promiseLoad)(ptr));
      }

      const children = new globalType.Lst();
      yield Promise.all(promiseList).then(values => {
        for (let i = 0; i < values.length; i++) {
          children.push(values[i]);
        }
      });
      return Promise.resolve(children);
    })();
  }
  /**
   * Returns the type of the relation
   * @return {Number} Type of the relation
   */


  getType() {
    return _SpinalRelationFactory.SPINAL_RELATION_LST_PTR_TYPE;
  }
  /**
   * Adds node as child of the relation. If the node is a model, creates a node.
   * @param node {SpinalNode | globalType.Model}
   */


  addChild(node) {
    if (node instanceof _SpinalNode.default) {
      this.children.push(new _SpinalNodePointer.default(node));
      this.getParent().then(parent => {
        if (typeof parent !== "undefined" && parent instanceof _SpinalNode.default) node._addParent(parent);
      });
    } else if (node instanceof globalType.Model) {
      const tmpNode = new _SpinalNode.default(this.name, node);
      this.addChild(tmpNode);
    }
  }
  /**
   * Removes a child from the relation.
   * @param {SpinalNode | Model} node Node to be removed 
   * @return {Boolean} true if the child has been successfully removed, false otherwise
   */


  removeChild(node) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      for (let i = 0; i < _this2.children.length; i++) {
        if (_this2.children[i].info.pointed_id === node.id) {
          let ptr = _this2.children[i];

          _this2.children.remove(ptr);
        }
      }

      return _this2.children.indexOf(node) === -1;
    })();
  }

}

_spinalCoreConnectorjs.default.register_models([SpinalRelationLstPtr]);

var _default = SpinalRelationLstPtr;
exports.default = _default;
