// spinalgraph/build/Relations/SpinalRelationPtrLst.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _BaseSpinalRelation = _interopRequireDefault(require("./BaseSpinalRelation"));

var _SpinalRelationFactory = require("./SpinalRelationFactory");

var _SpinalNode = _interopRequireDefault(require("../Nodes/SpinalNode"));

var _Utilities = require("../Utilities");

var _SpinalNodePointer = _interopRequireDefault(require("../SpinalNodePointer"));

var _spinalCoreConnectorjs = _interopRequireDefault(require("spinal-core-connectorjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

const globalType = typeof window === "undefined" ? global : window;

class SpinalRelationPtrLst extends _BaseSpinalRelation.default {
  constructor(name) {
    super(name);
    this.add_attr({
      children: new _SpinalNodePointer.default(new globalType.Lst())
    });
    this.children.info.add_attr("ids", []);
  }
  /**
   * This function retrieve all the id from children of the relation and return them inside an array.
   * @return {Array} containing all the children ids of the relation
   */


  getChildrenIds() {
    return this.children.ids;
  }
  /**
   * Return all the children of the relation
   * @return {Promise<globalType.Lst>}
   */


  getChildren() {
    return (0, _Utilities.promiseLoad)(this.children);
  }
  /**
   * Returns the type of the relation
   * @return {Number} Type of the relation
   */


  getType() {
    return _SpinalRelationFactory.SPINAL_RELATION_PTR_LST_TYPE;
  }
  /**
   * Add node as child of the relation if node is a model create a node
   * @param node {SpinalNode | globalType.Model}
   */


  addChild(node) {
    if (node instanceof _SpinalNode.default && !this.children.info.ids.contains(node.getId())) {
      this.children.info.ids.push(node.getId());
      (0, _Utilities.promiseLoad)(this.children).then(children => {
        children.push(node);
      });
      this.getParent().then(parent => {
        node._addParent(parent);
      });
    } else if (node instanceof globalType.Model) {
      const tmpNode = new _SpinalNode.default(this.name, node);
      this.addChild(tmpNode);
    }
  }
  /**
   * Remove the child from the relation
   * @param node {SpinalNode} child of the relation
   * @return {Promise<boolean>}
   */


  removeChild(node) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const childrenLst = yield (0, _Utilities.promiseLoad)(_this.children);
      childrenLst.remove(node);
      return Promise.resolve(childrenLst.indexOf(node) === -1);
    })();
  }

}

_spinalCoreConnectorjs.default.register_models([SpinalRelationPtrLst]);

var _default = SpinalRelationPtrLst;
exports.default = _default;
